torch/nn/utils/_expanded_weights/conv_expanded_weights.py

```diff
@@ -0,0 +1,67 @@
+import torch
+import torch.nn.functional as F
+
+from .conv_utils import conv_backward, conv_args_and_kwargs
+from .expanded_weights_impl import implements_per_sample_grads
+from .expanded_weights_utils import forward_helper
+
+THRESHOLD = 257
+
+@implements_per_sample_grads(F.conv1d)
+class Conv1dPerSampleGrad(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, kwarg_names, *expanded_args_and_kwargs):
+        if any([isinstance(i, str) for i in expanded_args_and_kwargs]):
+            raise RuntimeError("Expanded Weights does not support convolution padding as a string. "
+                               "Please file an issue to prioritize support")
+        expanded_args, expanded_kwargs = conv_args_and_kwargs(kwarg_names, expanded_args_and_kwargs)
+        output = forward_helper(F.conv1d, expanded_args, expanded_kwargs)
+        ctx.input, ctx.weight = expanded_args
+        ctx.bias = expanded_kwargs['bias']
+        ctx.stride, ctx.padding = expanded_kwargs['stride'], expanded_kwargs['padding']
+        ctx.dilation, ctx.groups = expanded_kwargs['dilation'], expanded_kwargs['groups']
+        return output
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        return conv_backward(F.conv1d, ctx, grad_output)
+
+
+@implements_per_sample_grads(F.conv2d)
+class Conv2dPerSampleGrad(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, kwarg_names, *expanded_args_and_kwargs):
+        if any([isinstance(i, str) for i in expanded_args_and_kwargs]):
+            raise RuntimeError("Expanded Weights does not support convolution padding as a string. "
+                               "Please file an issue to prioritize support")
+        expanded_args, expanded_kwargs = conv_args_and_kwargs(kwarg_names, expanded_args_and_kwargs)
+        output = forward_helper(F.conv2d, expanded_args, expanded_kwargs)
+        ctx.input, ctx.weight = expanded_args
+        ctx.bias = expanded_kwargs['bias']
```

Comment:
You don't need to save bias for backward -- you only need to save the shape of bias, right?

For ctx.input and ctx.weight -- we don't need to save input unless weight requires_grad. We don't need to save weight unless input requires grad.

These changes will help improve our memory utilization (I don't remember if Opacus does these already, if not, then all the better -- we're making a better system :) )

Reply:
Sadly with the current design we do need to save bias for the backward pass. Because expanded weights are producing gradients whose shape adds a batch dimension, I'm not having us use autograd directly but instead set the grad_sample attribute directly during the backward pass. Because it's setting the attribute directly, we needed to save the parameter

Reply:
Gotcha. So you need to save the weight, do you need to save the input? The input doesn't have a grad_sample field to populate right?

Reply:
Good point! Updated. Looks like the remaining functions after this (embedding + norms) call a backwards function that takes in input so I didn't update those